### PR TITLE
[SCB-925] Springmvc, when have defaultValue, required should be false, when param not exist, should check isRequired

### PR DESCRIPTION
**Pull request by weichao666**
…

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by weichao666** [comment]
Springmvc,
Generating schema：

if do not write defaultValue， required's value is what user write
if have written defaultValue，No matter how the user writes，required is false
running：

When param is not transmitted，if required is true, report error
When param is not transmitted , if required is false, handle it with defaultValue

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/CookieProcessorCreator.java [hunk 1]
```diff
@@ -35,33 +38,41 @@
...
-
+      if (value == null) {
+        value = checkRequiredAndDefaultValue(value);
+      }
       return convertValue(value, targetType);
     }
 
+    private Object checkRequiredAndDefaultValue(Object value) {
+      if (isRequired()) {
+        throw new InvocationException(Status.BAD_REQUEST, "Parameter is not valid, required is true");
```
Comment (liubao68) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/CookieProcessorCreator.java]: Parameter is reqired.

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/FormProcessorCreator.java [hunk 2]
```diff
@@ -50,20 +52,33 @@ public Object getValue(HttpServletRequest request) {
...
-        if (defaultValue != null) {
-          value = defaultValue;
-        }
+        value = checkRequiredAndDefaultValue(value);
       }
 
       return convertValue(value, targetType);
     }
 
+    private Object checkRequiredAndDefaultValue(Object values) {
```
Comment (liubao68) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/FormProcessorCreator.java]: values is always null, no need to pass this parameter

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/FormProcessorCreator.java [hunk 3]
```diff
@@ -50,20 +52,33 @@ public Object getValue(HttpServletRequest request) {
...
+      if (isRequired()) {
+        throw new InvocationException(Status.BAD_REQUEST, "Parameter is not valid, required is true");
+      }
+      Object defaultValue = getDefaultValue();
+      if (defaultValue != null) {
+        return defaultValue;
+      }
+      return values;
```
Comment (liubao68) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/FormProcessorCreator.java]: return getDefaultValue() is enouph

File: integration-tests/it-consumer/src/main/java/org/apache/servicecomb/it/testcase/TestDefaultValue.java [hunk 4]
```diff
@@ -472,12 +524,21 @@ public void floatHeader_require_springmvc_rt() {
...
+      assertEquals(true, e.getMessage().contains("Parameter is not valid"));
+    }
   }
 
   @Test
   public void floatForm_require_springmvc_rt() {
-    assertEquals(defaultFloat,
-        consumersSpringmvc.getSCBRestTemplate().postForObject("/floatFormRequire", null, float.class), 0.0f);
+    try {
+      consumersSpringmvc.getSCBRestTemplate().postForObject("/floatFormRequire", null, float.class);
```
Comment (liubao68) [inline integration-tests/it-consumer/src/main/java/org/apache/servicecomb/it/testcase/TestDefaultValue.java]: Is this behavior change? We need to notice in RN.

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/CookieProcessorCreator.java (hunk 1 above)
Comment (weichao666) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/CookieProcessorCreator.java]: done

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/FormProcessorCreator.java (hunk 2 above)
Comment (weichao666) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/FormProcessorCreator.java]: done

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/FormProcessorCreator.java (hunk 3 above)
Comment (weichao666) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/FormProcessorCreator.java]: done

File: integration-tests/it-consumer/src/main/java/org/apache/servicecomb/it/testcase/TestDefaultValue.java (hunk 4 above)
Comment (weichao666) [reply integration-tests/it-consumer/src/main/java/org/apache/servicecomb/it/testcase/TestDefaultValue.java]: yes

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/CookieProcessorCreator.java [hunk 5]
```diff
@@ -35,33 +38,37 @@
...
+      Object value = null;
+      if (cookies == null || cookies.length == 0) {
+        value = checkRequiredAndDefaultValue();
+        return convertValue(value, targetType);
       }
 
-      String value = null;
       for (Cookie cookie : cookies) {
         if (paramPath.equals(cookie.getName())) {
           value = cookie.getValue();
```
Comment (liubao68) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/CookieProcessorCreator.java]: It's better to add a break for this for statement

File: common/common-rest/src/test/java/org/apache/servicecomb/common/rest/codec/param/TestCookieProcessor.java [hunk 6]
```diff
@@ -67,8 +71,12 @@ public void testGetValueNoCookies() throws Exception {
     };
 
     CookieProcessor processor = createProcessor("c1", String.class);
-    Object value = processor.getValue(request);
-    Assert.assertNull(value);
+    try {
+      processor.getValue(request);
```
Comment (liubao68) [inline common/common-rest/src/test/java/org/apache/servicecomb/common/rest/codec/param/TestCookieProcessor.java]: This is a behavior change. And it's better to add a new test case which have the same meaning as the old test case. 
CookieProcessor processor = createProcessor("c1", String.class, null, false);

**Pull request merged**